File: client/go/main/udp.go
// Package main implements a UDP bridge client for X-Plane data references.
// This module provides functionality to communicate with X-Plane through UDP.
package main

import (
	"fmt"
	"net"
	"strings"
	"time"

	"github.com/fatih/color"
)

// UdpClient represents a UDP client for communicating with X-Plane.
// It encapsulates the connection details, server address, and timeout configuration
// for sending requests and receiving responses.
type UdpClient struct {
	serverAddr *net.UDPAddr  // The UDP address of the X-Plane server
	connection *net.UDPConn  // The UDP connection for communication
	timeout    time.Duration // Timeout duration for read operations
}

// NewUdpClient creates a new UDP client for communicating with X-Plane.
// It establishes a UDP connection to the specified server with the given timeout.
//
// Parameters:
//   - host: The IP address or hostname of the X-Plane server.
//   - port: The port number on which X-Plane is listening for UDP connections.
//   - timeoutSecs: The timeout in seconds for read operations.
//
// Returns:
//   - *UdpClient: A pointer to the newly created UDP client.
//   - nil: If the connection could not be established.
//
// Example:
//
//	client := NewUdpClient("127.0.0.1", 49000, 5)
//	if client != nil {
//	    // Use client for communication
//	}
func NewUdpClient(host string, port, timeoutSecs int) *UdpClient {
	fmt.Println(strings.Repeat("=", 100))
	color.Cyan("Creating UDP client to server %s:%d with timeout %d seconds\n", host, port, timeoutSecs)

	serverAddr := &net.UDPAddr{
		IP:   net.ParseIP(host),
		Port: port,
	}
	timeout := time.Duration(timeoutSecs) * time.Second

	connection, err := net.DialUDP("udp", nil, serverAddr)
	if err != nil {
		color.Red("UDP error while creating client: %v\n", err)
		return nil
	}

	color.Green("Created UDP client successfully\n")

	return &UdpClient{
		serverAddr: serverAddr,
		connection: connection,
		timeout:    timeout,
	}
}

// SendAndRecv sends data to the X-Plane server and waits for a response.
// It handles the complete request-response cycle with proper timeout handling.
//
// The method:
// 1. Sends the provided data to the connected server
// 2. Sets a read deadline based on the client's timeout configuration
// 3. Waits for a response up to the specified timeout
// 4. Returns the received data or nil if an error occurs
//
// Parameters:
//   - data: The byte slice containing the data to send to X-Plane.
//
// Returns:
//   - []byte: The response received from X-Plane.
//   - nil: If an error occurs during sending or receiving.
//
// Note:
//
//	The method uses a 2048-byte buffer for receiving data, which is
//	sufficient for typical X-Plane data reference responses.
func (client *UdpClient) SendAndRecv(data []byte) []byte {
	_, err := client.connection.Write(data)
	if err != nil {
		color.Red("UDP error while sending data: %v\n", err)
		return nil
	}

	err = client.connection.SetReadDeadline(time.Now().Add(client.timeout))
	if err != nil {
		color.Red("UDP error while setting read deadline: %v\n", err)
		return nil
	}

	buffer := make([]byte, 2048)

	size, _, err := client.connection.ReadFromUDP(buffer)
	if err != nil {
		color.Red("UDP error while reading data: %v\n", err)
		return nil
	}

	return buffer[:size]
}

File: client/rust/src/udp.rs
use std::io;
use std::io::ErrorKind;
use std::net::UdpSocket;
use std::time::Duration;

/// UDP client for communicating with the XPlane UDP bridge plugin.
///
/// This struct encapsulates the UDP socket and server address needed for
/// communication with the XPlane UDP bridge plugin. It provides basic
/// functionality for sending data and receiving responses.
pub(crate) struct UdpClient {
    /// Server address (e.g., "127.0.0.1:49000")
    server_addr: String,

    /// UDP socket for communication with the server
    socket: UdpSocket,
}

impl UdpClient {
    /// Creates a new UDP client instance.
    ///
    /// # Arguments
    ///
    /// * `host` - Server IP address (e.g., "127.0.0.1")
    /// * `port` - Server port (e.g., 49000)
    /// * `timeout_secs` - Socket timeout in seconds (e.g., 30)
    ///
    /// # Returns
    ///
    /// * `Ok(UdpClient)` - Successfully created client instance
    /// * `Err(io::Error)` - Error information if creation fails
    ///
    /// # Examples
    ///
    /// ```rust
    /// let client = UdpClient::new("127.0.0.1", 49000, 30)?;
    /// ```
    pub(crate) fn new(host: &str, port: u16, timeout_secs: u64) -> io::Result<Self> {
        println!("🔗 Connecting to {}:{} with timeout {} seconds", host, port, timeout_secs);

        let server_addr = format!("{}:{}", host, port);

        // Bind to local random port for client socket
        let socket = UdpSocket::bind("0.0.0.0:0")?;

        // Set socket read timeout
        socket.set_read_timeout(Some(Duration::from_secs(timeout_secs)))?;

        println!("✅  Connected successfully via UDP protocol");

        Ok(Self { server_addr, socket })
    }

    /// Sends data and waits for a response.
    ///
    /// # Arguments
    ///
    /// * `data` - Byte data to send
    ///
    /// # Returns
    ///
    /// * `Some(Vec<u8>)` - Response data received on success
    /// * `None` - Timeout or any error occurred
    ///
    /// # Examples
    ///
    /// ```rust
    /// let response = client.send_and_recv(&[0x01, 0x02, 0x03]);
    /// if let Some(data) = response {
    ///     println!("Received: {:?}", data);
    /// }
    /// ```
    pub(crate) fn send_and_recv(&self, data: &[u8]) -> Option<Vec<u8>> {
        // Send data
        if let Err(e) = self.socket.send_to(data, &self.server_addr) {
            eprintln!("❌ UDP error while sending data: {}", e);
            return None;
        }

        let mut buffer = [0u8; 2048];

        // Wait for UDP response
        match self.socket.recv_from(&mut buffer) {
            Ok((size, _src)) => Some(buffer[..size].to_vec()),
            Err(ref e) if e.kind() == ErrorKind::TimedOut => {
                let timeout = self.socket.read_timeout().unwrap().unwrap().as_secs();
                println!("⏰ UDP request timed out after {} seconds", timeout);
                None
            }
            Err(e) => {
                eprintln!("❌ UDP error while receiving data: {}", e);
                None
            }
        }
    }
}
